### PR TITLE
feat(map based calibrator): added 6dig search

### PR DESCRIPTION
**Pull request by YoheiMishina**
## Description
Functional modifications
Fixed to return the calibration result correctly.
Support for cases where calibration area map is not used.
Grid search for 6dig(added z, roll, pitch).

Improvement of launch
Calibration area map can be selected.
Online and offline can be selected.
Map based calibration configurations are added for each sensor model.

## Tests performed
test on offline with rosbag.
`ros2 launch extrinsic_calibration_manager calibration.launch.xml map_path:=/home/yohei/map/kashiwanoha pointcloud_map_file:=map.pcd lanelet2_map_file:=map.osm mode:=map_based sensor_model:=aip_xx1 vehicle_model:=jpntaxi use_calibration_area_map:=false `

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Comment by knzo25** [comment]
データお願いします！

**Comment by knzo25** [comment]
I tested the code with the data you provided me the other day in slack, but the calibration is failing.
Same as the test yesterday with the vehicle the map is not displayed (it is loaded though as the calibrator node is reading it. I also tested that I can get the map from ros2 topic).

The log is as follows:
```
[extrinsic_map_based_calibrator-29] [pcl::IterativeClosestPoint::computeTransformation] Not enough correspondences found. Relax your threshold parameters.
[extrinsic_map_based_calibrator-27] [pcl::IterativeClosestPoint::computeTransformation] Not enough correspondences found. Relax your threshold parameters.
[extrinsic_map_based_calibrator-33] [pcl::IterativeClosestPoint::computeTransformation] Not enough correspondences found. Relax your threshold parameters.
[extrinsic_map_based_calibrator-29] grid search time 9.957
[extrinsic_map_based_calibrator-29] [pcl::IterativeClosestPoint::computeTransformation] Not enough correspondences found. Relax your threshold parameters.
[extrinsic_calibration_manager-25] [INFO] [1657790674.577103768] [sensor_kit.extrinsic_calibration_manager]: Received service message: velodyne_right_base_link(success = 1 score = 5.40739e+09)
[extrinsic_calibration_manager-31] [INFO] [1657790676.164991674] [sensors.extrinsic_calibration_manager]: Waiting for responses...
[extrinsic_calibration_manager-25] [INFO] [1657790676.175110203] [sensor_kit.extrinsic_calibration_manager]: Waiting for responses...
[extrinsic_map_based_calibrator-27] grid search time 13.206
[extrinsic_map_based_calibrator-27] [pcl::IterativeClosestPoint::computeTransformation] Not enough correspondences found. Relax your threshold parameters.
[extrinsic_calibration_manager-25] [INFO] [1657790677.475050847] [sensor_kit.extrinsic_calibration_manager]: Received service message: velodyne_left_base_link(success = 1 score = 5.40619e+09)
[extrinsic_map_based_calibrator-33] grid search time 14.46
[extrinsic_map_based_calibrator-33] [pcl::IterativeClosestPoint::computeTransformation] Not enough correspondences found. Relax your threshold parameters.
[extrinsic_calibration_manager-31] [INFO] [1657790678.836296225] [sensors.extrinsic_calibration_manager]: Received service message: velodyne_rear_base_link(success = 1 score = 5.40683e+09)
```

It would seem the localization is failing? However I could not look into it as the map is not displayed :astonished:

**Comment by knzo25** [comment]
Unrelated to this error in particular, but for those kind of scores the success should be set to false should not it?

**Comment by YoheiMishina** [comment]
> Unrelated to this error in particular, but for those kind of scores the success should be set to false should not it?

I update judgement.
ec858291bcaccc8a9df2f6e0149f064e5cacf02c

**Comment by YoheiMishina** [comment]
> 

I changed parameter for using localization.
Cloud you try again?

**Comment by knzo25** [comment]
Regarding the [documentation](https://github.com/tier4/CalibrationTools/blob/tier4/universe/sensor/docs/how_to_extrinsic_map_based.md), its explanations about the setup are insufficient.

It should state, for example, which `tf` are required for the map_based method to work, which are either recorded in the rosbag or obtained via localization. Although we have more documentation in the private confluence, such information should also be here since we made the repository public.

Additionally, there is no much information about how the method works at the technical level, which is relevant for people interested in using the tool.

**Comment by knzo25** [comment]
> I update judgement.
https://github.com/tier4/CalibrationTools/commit/ec858291bcaccc8a9df2f6e0149f064e5cacf02c

Adding that default value for all calibration tools does not seem like a good idea.
For example, ICP/GICP do not always have that level of error (it would depend on the density of the map).
Additionally, the camera-lidar score is based on the reprojection error which is in pixel units (e.g., if the intrinsics are calibrated, a good error is less than 3-5 pixels depending on the camera resolution).

We should either set as a default value a quite relaxed one and then let each method choose their own in an optional fashion, or as you did use a very strict one as default, but look for a good value for all the implemented methods, because otherwise this PR breaks the other tools

**Comment by knzo25** [comment]
@YoheiMishina 
Did you fix the mismatch between the lidar frame (the one from which the data is from) and the lidar base frame (the one from the tf and the output yaml file)?

**Comment by YoheiMishina** [comment]
> @YoheiMishina Did you fix the mismatch between the lidar frame (the one from which the data is from) and the lidar base frame (the one from the tf and the output yaml file)?

Yes, I did fixed. But not include this PR.
And I will make next PR from [this branch](https://github.com/tier4/CalibrationTools/tree/feature/map_based_calib_update_ground_plane)

**Comment by YoheiMishina** [comment]
> > I update judgement.
> > [ec85829](https://github.com/tier4/CalibrationTools/commit/ec858291bcaccc8a9df2f6e0149f064e5cacf02c)
> 
> Adding that default value for all calibration tools does not seem like a good idea. For example, ICP/GICP do not always have that level of error (it would depend on the density of the map). Additionally, the camera-lidar score is based on the reprojection error which is in pixel units (e.g., if the intrinsics are calibrated, a good error is less than 3-5 pixels depending on the camera resolution).
> 
> We should either set as a default value a quite relaxed one and then let each method choose their own in an optional fashion, or as you did use a very strict one as default, but look for a good value for all the implemented methods, because otherwise this PR breaks the other tools

Changed default threshold to relaxed one. I forgot it was common threshold.
1113726e6ba684a6d3065119246fa727111f7a8e

File: sensor/extrinsic_map_based_calibrator/src/extrinsic_map_based_calibrator.cpp [hunk 1]
```diff
@@ -114,44 +124,59 @@ ExtrinsicMapBasedCalibrator::ExtrinsicMapBasedCalibrator(const rclcpp::NodeOptio
     calibrated_pointcloud_pub_ =
       this->create_publisher<sensor_msgs::msg::PointCloud2>("debug/calibrated_pointcloud", map_qos);
   }
+  // wait for other node
+  rclcpp::sleep_for(10s);
```
Comment (YoheiMishina) [inline sensor/extrinsic_map_based_calibrator/src/extrinsic_map_based_calibrator.cpp:128]: Timer to wait for Rviz to start.To be removed in the future.

**Comment by YoheiMishina** [comment]
7e0052781cab358f9d918204d45d7e2f50d1d437
Moved calling rviz node to map based calib.launch
This is because this Rviz is a tool that is included with map based calib.

**Review by knzo25 — APPROVED** [review]
LGTM

File: sensor/extrinsic_map_based_calibrator/src/extrinsic_map_based_calibrator.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (YoheiMishina) [reply sensor/extrinsic_map_based_calibrator/src/extrinsic_map_based_calibrator.cpp:128]: Timer to wait for Rviz to start.To be removed in the future.

**Pull request merged**